File: crates/compiler/frontend/typescript/src/workflow.ts
// One sealed typed Program contract beneath Workflow and Agent declarations.

import type { Json } from "./contract.js";
import type { ProgramBinding } from "./capture.js";
import { captureProgram } from "./capture.js";
import { authoredSource } from "./authored-source.js";
import { declaredSoFar, recordDeclaration } from "./declared.js";
import { compilerService } from "./compiler-service.js";
import { stableDigest } from "./markers.js";
import type { ContextSchema, ModelBinding } from "./markers.js";

// Capture host intrinsics before evaluated source runs.  The public handle is
// consumed after source evaluation, so calls through mutable global JSON/Object
// properties would otherwise let source replace the clone/freeze operations.
const jsonParse = JSON.parse.bind(JSON);
const jsonStringify = JSON.stringify.bind(JSON);
const objectFreeze = Object.freeze.bind(Object);
const objectValues = Object.values.bind(Object);
const WeakSetConstructor = WeakSet;

declare const workflowBrand: unique symbol;

/** A typed compiled program, including Context, loops, Hooks and yield/resume. */
export interface Program<Input, Output, Context = undefined> extends ProgramBinding {
  readonly [workflowBrand]: true;
  frontendGraph(): Json;
  diagnostics(): string | null;
  canonicalAir(): string;
  artifact(): Json;
  readonly new: (options?: { context?: Context }) => ProgramInstance<Input, Output, Context>;
  readonly invoke: (input: Input) => Promise<Output>;
}

export type ProgramConfig<Input, Output, Context> = {
  readonly name?: string;
  readonly context?: ContextSchema;
  run(agent: { context: Context; yield_(output: Output): Promise<Input> }, input: Input): Promise<Output> | Output;
};

/** General orchestration: Model calls are optional, never implicitly provided. */
export function Workflow<Input, Output, Context = undefined>(config: ProgramConfig<Input, Output, Context>): Program<Input, Output, Context> {
  return defineProgram(config, "workflow");
}

/** One capture boundary for both public declarations. */
export function defineProgram<Input, Output, Context>(
  config: ProgramConfig<Input, Output, Context>,
  declarationKind: "agent" | "workflow",
  primaryModel?: ModelBinding<never, unknown>,
): Program<Input, Output, Context> {
  const programId = config.name ?? (config.run.name || (declarationKind === "agent" ? "Agent" : "Workflow"));
  const graph = captureProgram({
    programId,
    entrypoint: programId,
    contextSchema: config.context,
    declarationKind,
    primaryModel,
    declared: declaredSoFar(),
    source: authoredSource(),
  });
  return recordDeclaration(new ProgramHandle<Input, Output, Context>(programId, graph));
}

/**
 * Keep captured graph data immutable while it crosses the evaluated-source
 * boundary.  A shallow `readonly` type is not a runtime security boundary:
 * submitted code can cast the graph to `any` and rewrite nested declarations
 * or requirements after capture.  The graph is JSON data, so recursively
 * freezing objects and arrays is both sufficient and deterministic.
 */
function freezeJson(value: Json): Json {
  const seen = new WeakSetConstructor<object>();

  const freeze = (current: unknown): void => {
    if (current === null || typeof current !== "object" || seen.has(current)) {
      return;
    }
    seen.add(current);
    for (const child of objectValues(current as Record<string, unknown>)) {
      freeze(child);
    }
    objectFreeze(current);
  };

  freeze(value);
  return value;
}

export class ProgramHandle<Input, Output, Context> implements ProgramBinding, Program<Input, Output, Context> {
  declare readonly [workflowBrand]: true;
  #graph: Json;
  readonly kind = "agent_definition" as const;
  readonly artifactDigest: string;
  readonly entrypoint: string;
  readonly targetAgentIdentityRequirement: string;

  constructor(
    readonly programId: string,
    graph: Json,
  ) {
    this.#graph = freezeJson(graph);
    this.artifactDigest = stableDigest(jsonStringify(this.#graph));
    this.entrypoint = programId;
    this.targetAgentIdentityRequirement = `${programId}.identity`;
    // `readonly` and `private` disappear at runtime. Freeze the handle as
    // well, so an `any` cast cannot replace its internal graph reference.
    objectFreeze(this);
  }

  frontendGraph(): Json {
    // Return a fresh value so source that inspects the graph cannot mutate the
    // value the capture harness requests after module evaluation.
    return jsonParse(jsonStringify(this.#graph)) as Json;
  }

  diagnostics(): string | null {
    return compilerService().verifyGraph(this.#graph);
  }

  canonicalAir(): string {
    return compilerService().canonicalAir(this.#graph);
  }

  artifact(): Json {
    return compilerService().artifact(this.#graph);
  }

  new(_options?: { context?: Context }): ProgramInstance<Input, Output, Context> {
    return new ProgramInstance<Input, Output, Context>(this.programId, this.artifactDigest);
  }

  invoke(_input: Input): Promise<Output> {
    throw new Error("Agent.invoke is called inside a compiled Agent body");
  }
}

// The evaluated source can otherwise replace a method on the shared
// prototype after an instance is created, bypassing the frozen instance.
objectFreeze(ProgramHandle.prototype);

export class ProgramInstance<Input, Output, Context> {
  constructor(
    readonly programRef: string,
    readonly artifactDigest: string,
  ) {}

  invoke(_input: Input): Promise<Output> {
    throw new Error("instance.invoke is called inside a compiled Agent body");
  }
}

File: crates/compiler/frontend/python/apxm_program/_workflow.py
"""Typed compiled Workflow handle shared by Agent authoring and composition."""

from __future__ import annotations

import hashlib
import json
from typing import Any, Awaitable, Callable, Generic, Optional, Protocol, TypeVar, overload

from . import _bridge, _native
from ._markers import ContextSchema

InputT = TypeVar("InputT")
OutputT = TypeVar("OutputT")
ContextT = TypeVar("ContextT")


class Program(Protocol[InputT, OutputT, ContextT]):
    """One compiled program, including Context, loops, Hooks and yield/resume."""

    @property
    def program_id(self) -> str: ...

    @property
    def _program_reference(self) -> tuple[str, str, str, str]: ...

    def frontend_graph(self) -> dict[str, Any]: ...
    def diagnostics(self) -> Optional[str]: ...
    def canonical_air(self) -> str: ...
    def artifact(self) -> dict[str, Any]: ...
    def new(self, *, context: Optional[ContextT] = None) -> ProgramInstance[InputT, OutputT, ContextT]: ...
    def invoke(self, _input: InputT) -> Awaitable[OutputT]: ...


@overload
def Workflow(*, input: type[InputT], output: type[OutputT], context: None = None) -> Callable[[Callable[..., Awaitable[OutputT]]], Program[InputT, OutputT, None]]: ...

@overload
def Workflow(*, input: type[InputT], output: type[OutputT], context: type[ContextT] | ContextSchema[ContextT]) -> Callable[[Callable[..., Awaitable[OutputT]]], Program[InputT, OutputT, ContextT]]: ...


def Workflow(
    *, input: type[InputT], output: type[OutputT],
    context: type[ContextT] | ContextSchema[ContextT] | None = None,
) -> Callable[[Callable[..., Awaitable[OutputT]]], Program[InputT, OutputT, ContextT]]:
    """General orchestration; Model calls are optional and never implicit."""
    from ._agent import _define_program

    return _define_program(input=input, output=output, context=context)


class _FrozenDefinitionType(type):
    """Prevent evaluated source from replacing Agent handle methods."""

    def __setattr__(cls, name: str, value: Any) -> None:
        if cls.__dict__.get("_methods_frozen", False):
            raise AttributeError("Workflow methods are immutable after import")
        super().__setattr__(name, value)


class ProgramHandle(Generic[InputT, OutputT, ContextT], metaclass=_FrozenDefinitionType):
    """A compiled typed Agent program and its composition surface."""

    __slots__ = (
        "_program_id",
        "_input_type_ref",
        "_output_type_ref",
        "_context_type_ref",
        "_artifact_digest",
        "_sealed",
        "__weakref__",
    )

    def __setattr__(self, name: str, value: Any) -> None:
        """Keep the captured handle's methods and snapshot reference stable."""
        if getattr(self, "_sealed", False):
            raise AttributeError("ProgramHandle is immutable after capture")
        object.__setattr__(self, name, value)

    def __init__(
        self,
        program_id: str,
        graph: dict[str, Any],
        input_type_ref: str,
        output_type_ref: str,
        context_type_ref: Optional[str],
    ) -> None:
        self._program_id = program_id
        # Keep the canonical snapshot in the native bridge. Python objects,
        # function defaults, and module globals are reflective from evaluated
        # source and therefore cannot serve as an integrity boundary.
        graph_json = json.dumps(graph, sort_keys=True, separators=(",", ":"))
        self._input_type_ref = input_type_ref
        self._output_type_ref = output_type_ref
        self._context_type_ref = context_type_ref
        self._artifact_digest = "sha256:" + hashlib.sha256(
            graph_json.encode("utf-8")
        ).hexdigest()
        _native.seal_frontend_graph(self, graph_json)
        self._sealed = True

    def _graph_value(self) -> dict[str, Any]:
        """Decode an independent graph value from the immutable snapshot."""
        return _native.frontend_graph(self)

    @property
    def program_id(self) -> str:
        return self._program_id

    @property
    def _program_reference(self) -> tuple[str, str, str, str]:
        """Return the static composition reference consumed by source capture."""
        return (
            self._program_id,
            self._artifact_digest,
            self._program_id,
            f"{self._program_id}.identity",
        )

    def frontend_graph(self) -> dict[str, Any]:
        """The captured FrontendGraph for this program."""
        return _native.frontend_graph(self)

    def diagnostics(self) -> Optional[str]:
        """Verification diagnostics for the captured graph, or ``None``."""
        return _bridge.verify_graph(self._graph_value())

    def canonical_air(self) -> str:
        """Canonical AIR JSON lowered from the captured graph."""
        return _bridge.canonical_air_json(self._graph_value())

    def artifact(self) -> dict[str, Any]:
        """One complete executable artifact compiled from the captured graph."""
        return _bridge.compile_artifact(self._graph_value())

    def new(self, *, context: Optional[ContextT] = None) -> "ProgramInstance[InputT, OutputT, ContextT]":
        """Create a stateful instance handle for later invocation."""
        return ProgramInstance(self._program_id, self)

    def invoke(self, _input: InputT) -> Awaitable[OutputT]:  # pragma: no cover
        raise RuntimeError("Agent.invoke is called inside a compiled Agent body")


type.__setattr__(ProgramHandle, "_methods_frozen", True)


class ProgramInstance(Generic[InputT, OutputT, ContextT]):
    """An inferred stateful instance handle produced by ``Agent.new``."""

    def __init__(self, program_ref: str, definition: Program[InputT, OutputT, ContextT]) -> None:
        self.program_ref = program_ref
        self._definition = definition

    def invoke(self, _input: InputT) -> Awaitable[OutputT]:  # pragma: no cover
        raise RuntimeError("instance.invoke is called inside a compiled Agent body")
